File: src/lib/ingest/meta.ts
/**
 * Repo metadata: `.frznforge.json` from the default branch tree (never the working copy),
 * merged with site-config overrides and derived defaults.
 */
import path from 'node:path';
import { RepoMetaInput, Slug, type RepoLinks, type Warning } from '../data/schema';
import { gitMaybe, readBlob } from './git';

export const META_FILENAME = '.frznforge.json';
export const MAX_DESCRIPTION = 300;

export interface RepoMetaFileResult {
  meta: RepoMetaInput | null;
  warnings: Warning[];
}

/** Truncate to MAX_DESCRIPTION chars (297 + "…"). Returns the input when short enough. */
export function truncateDescription(desc: string): string {
  const chars = Array.from(desc);
  if (chars.length <= MAX_DESCRIPTION) return desc;
  return chars.slice(0, MAX_DESCRIPTION - 3).join('') + '…';
}

/**
 * Read + validate `.frznforge.json` from `<treeish>:.frznforge.json`. Missing ⇒ meta null,
 * no warnings. Invalid JSON / schema ⇒ `repo-meta-invalid`, meta null. Over-long
 * description ⇒ truncated + `description-truncated`.
 */
export async function readRepoMetaFile(repo: string, treeish: string): Promise<RepoMetaFileResult> {
  const warnings: Warning[] = [];
  const sha = await gitMaybe(repo, ['rev-parse', '--verify', '--quiet', `${treeish}:${META_FILENAME}`]);
  if (sha === null || !sha.trim()) return { meta: null, warnings };
  const content = await readBlob(repo, sha.trim());
  let parsed: unknown;
  try {
    parsed = JSON.parse(content.toString('utf8'));
  } catch (e) {
    warnings.push({
      code: 'repo-meta-invalid',
      repo: null,
      message: `${META_FILENAME} is not valid JSON (${(e as Error).message}); ignored`,
    });
    return { meta: null, warnings };
  }
  if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
    const obj = parsed as Record<string, unknown>;
    if (typeof obj.description === 'string' && Array.from(obj.description).length > MAX_DESCRIPTION) {
      obj.description = truncateDescription(obj.description);
      warnings.push({
        code: 'description-truncated',
        repo: null,
        message: `${META_FILENAME} description exceeded ${MAX_DESCRIPTION} characters and was truncated`,
      });
    }
  }
  const result = RepoMetaInput.safeParse(parsed);
  if (!result.success) {
    const issues = result.error.issues.map((i) => `${i.path.join('.') || '(root)'}: ${i.message}`).join('; ');
    return {
      meta: null,
      warnings: [
        { code: 'repo-meta-invalid', repo: null, message: `${META_FILENAME} failed validation (${issues}); ignored` },
      ],
    };
  }
  return { meta: result.data, warnings };
}

export interface MergedMeta {
  name: string;
  description: string | null;
  links: RepoLinks;
  tags: string[];
  template: boolean;
  /** SPDX override, if any. */
  license: string | null;
  releaseMode: 'tags';
}

/** Site-config overrides > in-repo file > derived defaults. */
export function mergeMeta(
  defaults: { name: string },
  fileMeta: RepoMetaInput | null,
  overrides: RepoMetaInput | undefined,
): MergedMeta {
  if (overrides?.description !== undefined && Array.from(overrides.description).length > MAX_DESCRIPTION) {
    throw new Error(
      `config overrides.description for '${defaults.name}' is longer than ${MAX_DESCRIPTION} characters`,
    );
  }
  const pick = <K extends keyof RepoMetaInput>(k: K): RepoMetaInput[K] | undefined =>
    overrides?.[k] !== undefined ? overrides[k] : fileMeta?.[k];
  const links = pick('links') ?? {};
  return {
    name: pick('name') ?? defaults.name,
    description: pick('description') ?? null,
    links: sortedLinks(links),
    tags: [...(pick('tags') ?? [])],
    template: pick('template') ?? false,
    license: pick('license') ?? null,
    releaseMode: pick('releaseMode') ?? 'tags',
  };
}

/** Rebuild links in schema key order (determinism). */
function sortedLinks(links: RepoLinks): RepoLinks {
  const out: RepoLinks = {};
  if (links.homepage !== undefined) out.homepage = links.homepage;
  if (links.issues !== undefined) out.issues = links.issues;
  if (links.donations !== undefined) out.donations = links.donations;
  if (links.upstream !== undefined) out.upstream = links.upstream;
  return out;
}

/** Directory basename → slug: lowercase, non-[a-z0-9] runs → '-', trimmed. */
export function slugify(name: string): string {
  const s = name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
  return s.length ? s : 'repo';
}

/** Slug for a repo source: explicit slug, else slugified directory basename. */
export function slugFor(absPath: string, explicit?: string): string {
  const slug = explicit ?? slugify(repoBasename(absPath));
  const ok = Slug.safeParse(slug);
  if (!ok.success) throw new Error(`invalid slug ${JSON.stringify(slug)} for ${absPath}`);
  return slug;
}

/** Directory basename; for bare repos named `foo.git` the `.git` suffix is dropped. */
export function repoBasename(absPath: string): string {
  const base = path.basename(path.resolve(absPath));
  return base.endsWith('.git') && base.length > 4 ? base.slice(0, -4) : base;
}

File: internal/ingest/meta.go
package ingest

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"path/filepath"
	"regexp"
	"strings"
	"unicode/utf16"

	"frznforge/internal/config"
	"frznforge/internal/model"
)

// MetaFilename is the per-repo metadata file, read from the default branch's tree — never
// from the working copy, so an uncommitted edit cannot change the artifact.
const MetaFilename = ".frznforge.json"

// MaxDescription is the artifact's description limit, counted in UTF-16 code units.
const MaxDescription = 300

// isDescriptionTooLong counts UTF-16 code units, because that is what `z.string().max(300)`
// counts. Counting code points instead would let 300 emoji (600 units) past this check and
// then fail artifact validation, taking the whole build down over someone else's metadata.
func isDescriptionTooLong(desc string) bool {
	return utf16Len(desc) > MaxDescription
}

func utf16Len(s string) int {
	n := 0
	for _, r := range s {
		n++
		if r > 0xffff {
			n++
		}
	}
	return n
}

// truncateDescription cuts to MaxDescription code units (299 + "…"), never between the two
// halves of a surrogate pair — a lone surrogate is valid in a JS string but mojibake wherever
// it is rendered.
func truncateDescription(desc string) string {
	if !isDescriptionTooLong(desc) {
		return desc
	}
	units := utf16.Encode([]rune(desc))
	end := MaxDescription - 1
	if end-1 < len(units) {
		if lead := units[end-1]; lead >= 0xd800 && lead <= 0xdbff {
			end--
		}
	}
	return string(utf16.Decode(units[:end])) + "…"
}

// RepoMetaResult is a parsed .frznforge.json plus whatever went wrong reading it.
type RepoMetaResult struct {
	Meta     *config.RepoMetaInput
	Warnings []model.Warning
}

// ReadRepoMetaFile reads and validates `<treeish>:.frznforge.json`.
//
// Missing means no metadata and no warning — most repos have none. Invalid JSON or a value of
// the wrong type is a `repo-meta-invalid` warning and no metadata: a repo publishing broken
// metadata must not fail the build of a whole forge. An over-long description is truncated
// with a `description-truncated` warning rather than rejected, since the rest of the file is
// still good.
func ReadRepoMetaFile(ctx context.Context, repo, treeish string) (RepoMetaResult, error) {
	res := RepoMetaResult{Warnings: []model.Warning{}}
	out, ok, err := GitMaybe(ctx, repo, "rev-parse", "--verify", "--quiet", treeish+":"+MetaFilename)
	if err != nil {
		return res, err
	}
	sha := strings.TrimSpace(out)
	if !ok || sha == "" {
		return res, nil
	}
	content, err := ReadBlob(ctx, repo, sha)
	if err != nil {
		return res, err
	}

	var anyValue any
	if err := json.Unmarshal(content, &anyValue); err != nil {
		res.Warnings = append(res.Warnings, model.Warning{
			Code:    "repo-meta-invalid",
			Repo:    nil,
			Message: fmt.Sprintf("%s is not valid JSON (%s); ignored", MetaFilename, err),
		})
		return res, nil
	}

	invalid := func(issues string) RepoMetaResult {
		// Deliberately replaces the warning list rather than appending: a file that fails
		// validation contributes nothing, truncation warning included.
		return RepoMetaResult{Meta: nil, Warnings: []model.Warning{{
			Code:    "repo-meta-invalid",
			Repo:    nil,
			Message: fmt.Sprintf("%s failed validation (%s); ignored", MetaFilename, issues),
		}}}
	}

	var meta config.RepoMetaInput
	dec := json.NewDecoder(strings.NewReader(string(content)))
	if err := dec.Decode(&meta); err != nil {
		return invalid(err.Error()), nil
	}
	if meta.Description != nil && isDescriptionTooLong(*meta.Description) {
		truncated := truncateDescription(*meta.Description)
		meta.Description = &truncated
		res.Warnings = append(res.Warnings, model.Warning{
			Code:    "description-truncated",
			Repo:    nil,
			Message: fmt.Sprintf("%s description exceeded %d characters and was truncated", MetaFilename, MaxDescription),
		})
	}
	if issues := validateRepoMeta(&meta); issues != "" {
		return invalid(issues), nil
	}
	res.Meta = &meta
	return res, nil
}

// validateRepoMeta applies the rules the zod schema applies, returning them joined the way
// zod's issue list is joined. The wording is ours — zod's exact message text is not
// reproducible here — but the accept/reject decision is the same.
func validateRepoMeta(m *config.RepoMetaInput) string {
	var issues []string
	if m.Name != nil && *m.Name == "" {
		issues = append(issues, "name: Too small: expected string to have >=1 characters")
	}
	if m.License != nil && *m.License == "" {
		issues = append(issues, "license: Too small: expected string to have >=1 characters")
	}
	if m.ReleaseMode != nil && *m.ReleaseMode != "tags" && *m.ReleaseMode != "provider" {
		issues = append(issues, fmt.Sprintf("releaseMode: Invalid option: expected one of \"tags\"|\"provider\", received %q", *m.ReleaseMode))
	}
	for i, t := range m.Tags {
		if t == "" {
			issues = append(issues, fmt.Sprintf("tags.%d: Too small: expected string to have >=1 characters", i))
		}
	}
	if m.Links != nil {
		for _, l := range []struct {
			key   string
			value *string
		}{
			{"homepage", m.Links.Homepage},
			{"issues", m.Links.Issues},
			{"donations", m.Links.Donations},
			{"upstream", m.Links.Upstream},
		} {
			if l.value == nil {
				continue
			}
			if u, err := url.Parse(*l.value); err != nil || u.Scheme == "" {
				issues = append(issues, "links."+l.key+": Invalid URL")
			}
		}
	}
	return strings.Join(issues, "; ")
}

// mergedMeta is the metadata the artifact actually records, after every layer has had its say.
type mergedMeta struct {
	Name        string
	Description *string
	Links       model.RepoLinks
	Tags        []string
	Template    bool
	// License is an SPDX override, "" when none was configured.
	License     string
	ReleaseMode string
}

// mergeMeta layers site-config overrides over the in-repo file over derived defaults.
func mergeMeta(defaultName string, fileMeta, overrides *config.RepoMetaInput) (mergedMeta, error) {
	if overrides != nil && overrides.Description != nil && isDescriptionTooLong(*overrides.Description) {
		return mergedMeta{}, fmt.Errorf(
			"config overrides.description for '%s' is longer than %d characters", defaultName, MaxDescription)
	}
	pickString := func(get func(*config.RepoMetaInput) *string) *string {
		if overrides != nil {
			if v := get(overrides); v != nil {
				return v
			}
		}
		if fileMeta != nil {
			return get(fileMeta)
		}
		return nil
	}

	out := mergedMeta{Name: defaultName, Tags: []string{}, ReleaseMode: "tags"}
	if v := pickString(func(m *config.RepoMetaInput) *string { return m.Name }); v != nil {
		out.Name = *v
	}
	out.Description = pickString(func(m *config.RepoMetaInput) *string { return m.Description })
	if v := pickString(func(m *config.RepoMetaInput) *string { return m.License }); v != nil {
		out.License = *v
	}
	if v := pickString(func(m *config.RepoMetaInput) *string { return m.ReleaseMode }); v != nil {
		out.ReleaseMode = *v
	}
	if overrides != nil && overrides.Tags != nil {
		out.Tags = append([]string{}, overrides.Tags...)
	} else if fileMeta != nil && fileMeta.Tags != nil {
		out.Tags = append([]string{}, fileMeta.Tags...)
	}
	if overrides != nil && overrides.Template != nil {
		out.Template = *overrides.Template
	} else if fileMeta != nil && fileMeta.Template != nil {
		out.Template = *fileMeta.Template
	}
	// links merge per key rather than being picked wholesale, which is what every other field
	// does: adding a `donations` link in the config must not delete the homepage the repo's own
	// file declared.
	out.Links = mergeLinks(linksOf(fileMeta), linksOf(overrides))
	return out, nil
}

func linksOf(m *config.RepoMetaInput) *config.RepoLinks {
	if m == nil {
		return nil
	}
	return m.Links
}

// mergeLinks builds the artifact links in schema key order — the key order IS the artifact's
// byte order, so it is rebuilt rather than copied.
func mergeLinks(low, high *config.RepoLinks) model.RepoLinks {
	pick := func(get func(*config.RepoLinks) *string) *string {
		if high != nil {
			if v := get(high); v != nil {
				return v
			}
		}
		if low != nil {
			return get(low)
		}
		return nil
	}
	return model.RepoLinks{
		Homepage:  pick(func(l *config.RepoLinks) *string { return l.Homepage }),
		Issues:    pick(func(l *config.RepoLinks) *string { return l.Issues }),
		Donations: pick(func(l *config.RepoLinks) *string { return l.Donations }),
		Upstream:  pick(func(l *config.RepoLinks) *string { return l.Upstream }),
	}
}

var (
	nonSlugRun = regexp.MustCompile(`[^a-z0-9]+`)
	slugRe     = regexp.MustCompile(`^[a-z0-9][a-z0-9-]*$`)
)

// Slugify turns a directory basename into a slug: lowercase, runs of anything else to '-',
// trimmed. An empty result becomes "repo" rather than an unusable empty path segment.
func Slugify(name string) string {
	s := nonSlugRun.ReplaceAllString(strings.ToLower(name), "-")
	s = strings.Trim(s, "-")
	if s == "" {
		return "repo"
	}
	return s
}

// SlugFor is a repo source's slug: the explicit one, else the slugified directory basename.
func SlugFor(absPath, explicit string) (string, error) {
	slug := explicit
	if slug == "" {
		slug = Slugify(RepoBasename(absPath))
	}
	if !slugRe.MatchString(slug) {
		return "", fmt.Errorf("invalid slug %q for %s", slug, absPath)
	}
	return slug, nil
}

// RepoBasename is the directory name; a bare repo named `foo.git` reports as `foo`.
func RepoBasename(absPath string) string {
	abs, err := filepath.Abs(absPath)
	if err != nil {
		abs = absPath
	}
	base := filepath.Base(abs)
	if strings.HasSuffix(base, ".git") && len(base) > 4 {
		return base[:len(base)-4]
	}
	return base
}
